pkg/controller/sparkapplication/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,162 @@
+package volcano
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"volcano.sh/volcano/pkg/apis/scheduling/v1alpha2"
+	volcanoclient "volcano.sh/volcano/pkg/client/clientset/versioned"
+
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta1"
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/controller/sparkapplication/batchscheduler/interface"
+)
+
+const (
+	PodGroupName = "podgroups.scheduling.incubator.k8s.io"
+)
+
+type VolcanoBatchScheduler struct {
+	extensionClient apiextensionsclient.Interface
+	volcanoClient   volcanoclient.Interface
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string) bool {
+		return scheduler != nil && *scheduler == v.Name()
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName) && checkScheduler(app.Spec.Driver.SchedulerName)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) OnSubmitSparkApplication(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	if app.Spec.Executor.Annotations == nil {
+		app.Spec.Executor.Annotations = make(map[string]string)
+	}
+
+	if app.Spec.Driver.Annotations == nil {
+		app.Spec.Driver.Annotations = make(map[string]string)
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupForClientAPP(app)
+	} else if app.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupForClusterAPP(app)
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClientAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	if _, ok := app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, *app.Spec.Executor.Instances); err == nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClusterAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We need both mark Driver and Executor when submitting
```

Comment:
The comment should explain why the group size is initially 1 and how it's adjusted once the driver pod gets scheduled.